wellcomecollection/editorial-photography-ingest
# Parallel downloads

## What does this change?

This adds the ability to pull multiple S3 objects in parallel.

resolves https://github.com/wellcomecollection/editorial-photography-ingest/issues/11

## How to test
Using an AWS profile with permission to download from  wellcomecollection-editorial-photography...

`PYTHONPATH=src python src/transferrer/download.py PBBD_TEST`

The log output should show two `downloading` lines, followed by two `downloaded` lines

```
INFO:__main__:downloading	ST/PB_BD_TEST/PB_TEST_001.tif
INFO:__main__:downloading	ST/PB_BD_TEST/PB_TEST_002.tif
INFO:__main__:downloaded	ST/PB_BD_TEST/PB_TEST_001.tif
INFO:__main__:downloaded	ST/PB_BD_TEST/PB_TEST_002.tif
```

## How can we measure success?

Downloads of s3 folders should be faster.  However, this is a premature optimisation, as I don't have a real-world (i.e. running on Lambda) example to compare it against.  Running on my computer, fetching EPOPTEST (not currently available), it seemed to shave off a few seconds.

Even so, I expect this to make it possible for us to tune the performance of the process.


## Have we considered potential risks?

This is a little bit of added complexity, which has the potential to cause the process to use significantly more memory and possibly fail.  In order to mitigate this, the thread pool is constrained to a relatively low number (0.01 of the max threads available in Lambda)
